### ISSUE TITLE
rds: possible LDS+RDS race condition ends in unknown behavior.

### ISSUE THREAD

**Issue by stevenzzzz** [envoyproxy/envoy#7902]
**Bug Template**

*Title*: Fast LDS tearing down a HCM may interact with RDS update callbacks on the fly in worker threads.

*Description*:
>IIUC, there is a race condition that a RDS update propagation on worker threads may be interfered by the destruction of the RdsRouteConfigProvider on the main thread, which ends in seg-fault or other unknown behaviors. 

*Repro steps*:
> 
- RDS propagates a configUpdate to workers, the posted callback is as following:
tls_->runOnAllThreads(
  [this, new_config]() -> void { tls_->getTyped<ThreadLocalConfig>().config_ = new_config; }
}
- LDS updates the active listener in main thread, tear down the RdsRouteConfigProvider, the "this" above is released.
- A callback gets executed in a worker thread, ends in a seg-fault or a invalid address access.

I have some possible fix for this if this is a real issue.

**Comment by stevenzzzz** [comment]
@mattklein123 is this an real issue?

**Comment by mattklein123** [comment]
Yes definitely a real bug. I think there are likely multiple ways to fix but I would have to dig in myself to have a recommendation so maybe just propose a fix and we can discuss in code review?

**Comment by stevenzzzz** [comment]
A quick solution is to "relax" the ownership of RdsRouteConfigProvider a little bit at the HCM, holds an shared_ptr instead of a unique_ptr of the RdsRouteConfigProvider, when we do onConfigUpdate(), capture weak_ptr of the RdsRouteConfigProvider, and in the update callback, check if we can lock a shared_ptr of the provider before we apply the change.

The complication of this change is that the destruction of RdsRouteConfigProvider has to happen in the main thread, as it contains a tls_ memeber.

For that, I am gonna propose another slot-impl wrapper (or update the current impl), such that on destruction, it make sure the slotImpl is done in main thread. (another relaxation?) 

I don't know if this/these relaxation(s) is welcomed or not. 
But noteworthy SRDS will hold shared RdsRouteConfigProvider in threadlocal ScopedConfigImpl.

**Comment by mattklein123** [comment]
I chatted with @stevenzzzz offline, and my high level view of this is:
1) The optimal solution would be to cancel/kill all watches/subscriptions from the listener prior to commencing draining. This will work because draining itself is serialized with the workers. Before the listener is destroyed, all messages will have been flushed on the workers, so if the subscriptions are killed, there should be no race condition. The issue here is how much effort it will be to figure out a clean way to cancel watches/subscriptions outside of the context of the destructor. This will need investigation.
2) If (1) doesn't work, I do think that some type of weak pointer and/or relaxing where things can get deleted approach would work, but I suspect it will be subtly complicated and personally I would try for (1) first.

@stevenzzzz do you maybe want to investigate both (1) and (2) and come back with a proposal?

**Comment by stevenzzzz** [comment]
update to stop this been closed by stale bot.

This problem also applies to shutdown time, probably workers should drain their dispatcher queue.

**Comment by stevenzzzz** [comment]
Added a doc with some options I can think of, let's continue our discussion there.

https://docs.google.com/document/d/1_axmNmbkDO3aQE3jS5-WgSLvXyVxKwapOnrkBZn-DSk/edit?usp=sharing

**Comment by stevenzzzz** [comment]
I created a concept-proof PR to show how the detachable SlotImpl would be like,
PTAL

https://github.com/stevenzzzz/envoy/pull/11

**Comment by stevenzzzz** [comment]
I believe this can be closed now.
